instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/field/VirtualField.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.field;
+
+import io.opentelemetry.instrumentation.api.internal.RuntimeVirtualFieldSupplier;
+import java.util.function.Supplier;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Represents a "virtual" field of type {@code F} that is added to type {@code T} in the runtime.
+ *
+ * <p>Field values are weakly referenced and will be garbage collected when their owner instance is
```

Comment:
My understanding is that field values are not weakly referenced, if they were they could be collected if nobody else has a reference to them.
If I understand this correctly the happy path is that we transform the class and add a field to it so we really have a regular strong reference to the value. Things get messy when class is not transformed for some reason. Then we put  the values into a weak map where key is the object on which field is set. This has the annoying property that when value references key then we immediately have a memory leak because now key can be never collected. We could get around this by also making values weak which would mean that something else needs to keep the values alive. Basically afaik no matter what we do we can't get the same behaviour as when the value is just stored in a field.
Perhaps we could address this somehow. Maybe we could drop the support for using fallback map? Now that lambda classes are also transformed is there a case where it could be needed? There is a flag to disable field injection and there is hack that skips filed injection on classes annotated with `javax.decorator.Decorator` so might not be that easy to completely get rid of it. An alternative would be to have something in `VirtualField` api that lets user figure out whether it is field or map based. Or could add methods like `setWeak` that doesn't change anything in field backed implementation but for map based implementation makes the value also weak. This would allow using `VirtualField` in netty `FutureListenerWrappers` without leaking memory even when map based store is used.

Reply:
> My understanding is that field values are not weakly referenced, if they were they could be collected if nobody else has a reference to them.

Yeah that's true, I copied this Javadoc and didn't really think too deeply about it.

> Basically afaik no matter what we do we can't get the same behaviour as when the value is just stored in a field.
> Perhaps we could address this somehow. Maybe we could drop the support for using fallback map?

We probably have to keep the fallback implementation; I suppose it's better to have it in place and install a few more instrumentations than not to have it and skip instrumenting some pieces of code.

Perhaps we could fix the Javadoc so that it outright says that this has similar semantics to a weak-keys strong-values map and keeping values that might reference the original key is discouraged?

Reply:
> Maybe we could drop the support for using fallback map?

that would be nice, the memory leak has bit us before (at least #441 and #3027 I think)

> Now that lambda classes are also transformed is there a case where it could be needed? There is a flag to disable field injection and there is hack that skips filed injection on classes annotated with javax.decorator.Decorator so might not be that easy to completely get rid of it

I wonder if the `javax.decorator.Decorator` is needed since #3948

Reply:
This is the original DD PR that introduced this mathcer: https://github.com/DataDog/dd-trace-java/pull/1045

> In CDI (Contexts and Dependency Injection) environments, classes annotated with `@Decorator` wrap other classes.
Decorators implement the same interfaces but delegate to another class. See the Oracle docs.
> 
> In general, the tracer should instrument the delegate (decorated class) and not the decorator. When the tracer instruments the decorator, it cause the CDI container to be confused about the interfaces of the decorator and crash on startup.

If we can reproduce that problem and verify that #3948 solves it then we can start thinking of phasing out the weak map implementation.

Reply:
opened #4241 to track/discuss further